### PR TITLE
[query] frozendict

### PR DESCRIPTION
**Pull request by johnc1231**
CHANGELOG: Evaluating hail expressions in python will now return `frozendict`, an immutable dictionary type. Sets will return python's `frozenset`.

This is necessary because hail supports hashable dicts, but python does not. Same with sets.

### CODE REVIEW

**Review by tpoterba — CHANGES_REQUESTED** [review]
can you add a test to ensure we can correctly impute the type of these when we do `hl.literal()` or something?

File: hail/python/hail/utils/frozendict.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+from typing import TypeVar, Dict
+from collections.abc import Mapping
+
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+class frozendict(Mapping):
```
Comment (danking) [inline hail/python/hail/utils/frozendict.py:8]: I think Mapping can take type arguments `Mapping[T, U]`.

**Review by danking — CHANGES_REQUESTED** [review]

File: hail/python/hail/utils/frozendict.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        return self.d[k]
+
+    def __hash__(self) -> int:
+        return hash(frozenset(self.items()))
+
+    def __len__(self) -> int:
+        return len(self.d)
+
+    def __iter__(self):
+        return iter(self.d)
```
Comment (danking) [inline hail/python/hail/utils/frozendict.py:37]: Can you add a repr:
```
def __repr__(self): 
    return f'frozendict({self.d!r})'                                                                                 
```

File: hail/python/hail/utils/frozendict.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+T = TypeVar("T")
+U = TypeVar("U")
+
+
+class frozendict(Mapping):
+    """
+    An object representing an immutable dictionary.
+
+    >>> my_frozen_dict = hl.utils.frozendict({1:2, 7:5})
+
```
Comment (danking) [inline hail/python/hail/utils/frozendict.py:13]: Can you give an example of converting a frozendict back to a normal dict?

```
dict(frozendict({'a': 1, 'b': 2}))
```

**Review by danking — CHANGES_REQUESTED** [review]

File: hail/python/hail/utils/frozendict.py (hunk 1 above)
Comment (johnc1231) [reply hail/python/hail/utils/frozendict.py:8]: `typing.Dict` supports this, but `typing.Mapping` doesn't seem to, unless I'm misreading. I'm new to the python type annotations stuff.

**Comment by johnc1231** [comment]
@danking  @tpoterba  Done everything you both asked for.

**Review by tpoterba — APPROVED** [review]

**Pull request merged**